### PR TITLE
MINOR: [Go][CI] Add openssl to macos CGO runs

### PR DESCRIPTION
**Pull request by zeroshade**
### Rationale for this change
Fix the failing CI runs

### CODE REVIEW

File: .github/workflows/go.yml [hunk 1]
```diff
@@ -300,18 +300,22 @@ jobs:
         with:
           go-version: ${{ matrix.go }}
           cache: true
-          cache-dependency-path: go/go.sum
+          cache-dependency-path: go/go.sum
```
Comment (assignUser) [inline .github/workflows/go.yml:303]: nit:
```suggestion
          cache-dependency-path: go/go.sum
```

File: .github/workflows/go.yml [hunk 2]
```diff
@@ -300,18 +300,22 @@ jobs:
...
       - name: Brew Install Arrow and pkg-config
         shell: bash
         run: brew install apache-arrow pkg-config
       - name: Install staticcheck
         run: go install honnef.co/go/tools/cmd/staticcheck@${{ matrix.staticcheck }}
       - name: Build
         shell: bash
-        run: ci/scripts/go_build.sh $(pwd)
+        run: |
+          export PKG_CONFIG_PATH=/usr/local/Cellar/openssl@3/3.0.8/lib/pkgconfig:$PKG_CONFIG_PATH
```
Comment (assignUser) [inline .github/workflows/go.yml:312]: This should update it for all following steps by setting it once :)
```suggestion
           echo "PKG_CONFIG_PATH=/usr/local/Cellar/openssl@3/3.0.8/lib/pkgconfig:$PKG_CONFIG_PATH" >> $GITHUB_ENV
```

Comment (lidavidm) [inline .github/workflows/go.yml:312]: Hmm, hardcoding the path like this seems brittle. Are we missing a `brew link` or something similar?

File: .github/workflows/go.yml (hunk 1 above)
Comment (lidavidm) [inline .github/workflows/go.yml:303]: nit: trailing whitespace

File: .github/workflows/go.yml [hunk 3]
```diff
@@ -300,18 +300,22 @@ jobs:
...
+          ci/scripts/go_build.sh $(pwd)
       - name: Test
         shell: bash
-        run: ci/scripts/go_test.sh $(pwd)
+        run: |
+          export PKG_CONFIG_PATH=/usr/local/Cellar/openssl@3/3.0.8/lib/pkgconfig:$PKG_CONFIG_PATH
+          ci/scripts/go_test.sh $(pwd)
```
Comment (assignUser) [inline .github/workflows/go.yml:318]:
```suggestion
        run: ci/scripts/go_test.sh $(pwd)
```

File: .github/workflows/go.yml (hunk 2 above)
Comment (lidavidm) [reply .github/workflows/go.yml:312]: Ah, OpenSSL is special in that regard for Homebrew...

**Review by assignUser — COMMENTED** [review]
A minor change allows us to not have to export PKG_CONFIG_PATH in each step.

File: .github/workflows/go.yml (hunk 2 above)
Comment (zeroshade) [reply .github/workflows/go.yml:312]: From what I found, it seems like it's explicitly an issue with the openssl brew package not placing `openssl.pc` in the same spot as other pkg-config paths. :( If there's a `brew link` or something that we can do, I haven't been able to find it.

Comment (assignUser) [reply .github/workflows/go.yml]: IIRC we have something similar hardcoded in a bunch of places because there is no better solution?

Comment (lidavidm) [reply .github/workflows/go.yml]: Yeah, that makes sense. Oh well...

**Review by lidavidm — APPROVED** [review]

File: .github/workflows/go.yml [hunk 4]
```diff
@@ -308,7 +308,9 @@ jobs:
         run: go install honnef.co/go/tools/cmd/staticcheck@${{ matrix.staticcheck }}
       - name: Build
         shell: bash
-        run: ci/scripts/go_build.sh $(pwd)
+        run: |
+           echo "PKG_CONFIG_PATH=/usr/local/Cellar/openssl@3/3.0.8/lib/pkgconfig:$PKG_CONFIG_PATH" >> $GITHUB_ENV
```
Comment (kou) [inline .github/workflows/go.yml:312]: This doesn't affect in this step.

We need a separated step:

```yaml
- name: Prepare PKG_CONFIG_PATH
  shell: bash
  run: |
    echo "PKG_CONFIG_PATH=$(brew --prefix openssl)/lib/pkgconfig:$PKG_CONFIG_PATH" >> $GITHUB_ENV
- name: Build
  shell: bash
  run: ci/scripts/go_build.sh $(pwd)
...
```

And this change causes YAML syntax error: https://github.com/apache/arrow/actions/runs/4358242702/workflow#L308

I think that the indentation of this line is a problem.

Comment (assignUser) [reply .github/workflows/go.yml:312]: Also weird how the workflow didn't run on the last commit. Sorry for the indentation error that happens due to my comment suggestion 🙈

**Pull request merged**